### PR TITLE
Pull in existing code from talsma-ict.

### PR DESCRIPTION
**Pull request by sjoerdtalsma**
The java-globaltracer library has been split up in separate packages for _active span_ management and _global tracer_ access.

This pull request is for the _active span_ management package into an independent separate library.

### CODE REVIEW

**Comment by sjoerdtalsma** [comment]
@adriancole @wu-sheng I can't select you as reviewers, but I know you have a strong opinion about this as well 😄

**Comment by codefromthecrypt** [comment]
first peek is deja-vu from java-globaltracer.. either that's good or not :) I will take a look at this tomorrow after feedback from the former is digested

File: src/main/java/io/opentracing/contrib/activespan/concurrent/CallableWithActiveSpan.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.util.concurrent.Callable;
+
+/**
+ * {@link Callable} wrapper that will execute with the {@link ActiveSpanManager#activeSpan() active span}
+ * from the scheduling thread.
+ *
+ * @author Sjoerd Talsma
+ */
+public class CallableWithActiveSpan<T> implements Callable<T> {
```
Comment (codefromthecrypt) [inline src/main/java/io/opentracing/contrib/activespan/concurrent/CallableWithActiveSpan.java]: does this need to be a public type? similar for the Runnable variant

File: src/main/java/io/opentracing/contrib/activespan/ActiveSpanManager.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+     * @return The object that will restore any currently <em>active</em> deactivated, or <code>null</code>.
+     * @see #activeSpan()
+     * @see #deactivate(SpanDeactivator)
+     */
+    public static SpanDeactivator activate(Span span) {
+        try {
+            if (span == null) span = NoopSpan.INSTANCE;
+            return getInstance().setActiveSpan(span);
+        } catch (Exception activationException) {
+            LOGGER.log(Level.WARNING, "Could not activate {0}.", new Object[]{span, activationException});
```
Comment (codefromthecrypt) [inline src/main/java/io/opentracing/contrib/activespan/ActiveSpanManager.java]: seems string cat is simpler here. there's questionable value in adding a period vs the cruft. ps did you mean to add the exception as a parameter?

File: src/main/java/io/opentracing/contrib/activespan/ActiveSpanManager.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+ * @navassoc - activeSpan - io.opentracing.Span
+ */
+public abstract class ActiveSpanManager {
+    private static final Logger LOGGER = Logger.getLogger(ActiveSpanManager.class.getName());
+    private static final AtomicReference<ActiveSpanManager> INSTANCE = new AtomicReference<ActiveSpanManager>();
+
+    /**
+     * Managed object to deactivate an activated {@link Span} with.
+     */
+    public interface SpanDeactivator {
```
Comment (codefromthecrypt) [inline src/main/java/io/opentracing/contrib/activespan/ActiveSpanManager.java]: why the marker interface. what does one expect to do with it?

**Review by codefromthecrypt — COMMENTED** [review]
Similar to the other codebase I think this overshoots in complexity, which is particularly troubling as tests are absent. Please try to pare down to minimum features needed and don't add forward thinking extension points or expose anything public that isn't proven necessary by people besides yourself.

**Comment by codefromthecrypt** [comment]
the main thing I think you lose track of is that you are thinking somewhere possibly farther than others. It is possible that the end design will eventually become as complex, but since this is intended to be a shared project, I think it is better to focus code on only what's required and no more. Then, use things like the rule-of-three before making things public etc.

Especially when doing fancy code like the lock-free loops, you really must have these things tested as this is the sort of code that breaks and is hard to tell why. Tests are a great way to tell if code should be present or not. For example, if you feel strongly that some hook is necessary, a feature test can explain that better than anyone's conjecture of importance.

**Comment by sjoerdtalsma** [comment]
@bensigelman  Should we rename the repo to `java-spanmanager` in line with the new concepts?

**Comment by codefromthecrypt** [comment]
just to set expectations, I'm not planning to do further reviews on this, as between the global tracer and here, I've spent dozens of hours which delayed very important work I'm behind on. I hope my help so far has been.. helpful.

**Comment by bensigelman** [comment]
@adriancole your help has been helpful and your message about future help is also helpful. I just wanted to give you the courtesy of a final look if you wanted it, opt-in. Gracias.

File: src/main/java/io/opentracing/contrib/spanmanager/ThreadLocalSpanManager.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+ * <p>
+ * The linked managed spans provide the following stack unwinding algorithm:
+ * <ol>
+ * <li>If the released <em>span</em> is not the managed span, the managed span is left alone.</li>
+ * <li>Otherwise, the first parent that is <em>not yet released</em> is set as the new managed span.</li>
+ * <li>If no managed parents remain, the <em>managed span</em> is cleared.</li>
+ * <li>Consecutive <code>release()</code> calls for already-released spans will be ignored.</li>
+ * </ol>
+ */
+final class ThreadLocalSpanManager implements SpanManager {
```
Comment (bhs) [inline src/main/java/io/opentracing/contrib/spanmanager/ThreadLocalSpanManager.java]: Do you have any practical examples of possible non-TLS SpanManagers that would actually play nice with the static-ness of this API? (I.e., that could retrieve the active span *without parameters* and without using TLS) Or is the idea just that they'd use some *other* TLS, like MDC or similar?

File: README.md [hunk 5]
```diff
@@ -1 +1,57 @@
...
+The _active_ SpanManager is resolved as follows:
+ 1. The last-registered span manager always takes precedence.
+ 2. If no manager was registered, one is looked up from the `ServiceLoader`.  
+    The ActiveSpanManager will not attempt to choose between implementations:
+ 3. If no single implementation is found, the default SpanManager will be used.
+
+
+## Examples
+
+_TODO create examples_
```
Comment (bhs) [inline README.md]: This is arguably the most important / first thing we should do (even if there's no impl to back things up). It'd be great to talk about how the calling code is going to look and just collectively imagine impls, at least as long as we are able to imagine the same things.

**Review by bhs — COMMENTED** [review]
@sjoerdtalsma I want to make progress on this PR in the coming days if you have cycles for it.

FWIW, I would like OpenTracing APIs to change in 2017 in order to make things like this repo easier/better/cleaner. It's possible that something like this would actually get folded into OpenTracing proper, actually. I figured that getting into the weeds on this PR would be a good way to bring all of these issues to the fore.

I also wanted to mention an alternative: make a version number like `0.0.1`, add a giant caveat to the top of the README, and just clean things up a little bit and merge something that's a working prototype more than a fully-debated PR. We would then look at how this works when integrated into real/mature repos and make changes as needed.

Thanks!

File: src/main/java/io/opentracing/contrib/spanmanager/ThreadLocalSpanManager.java (hunk 4 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/ThreadLocalSpanManager.java]: Good question. The idea is that people can keep using their own context propagation mechanism they may already have, as that has impact on the code.
For instance, I created a security context propagation for a client that transfers to new background threads. I can easily create a SpanManager adaptor using that same propagation tool to _also_ propagate/piggyback the spans besides other contextual information.

The impact is on the code scheduling the new threads which needs to be decoupled from whatever context must be propagated.
Whatever works for somebody should be extendable to also propagate spans with, hence the SpanManager interface.

I agree there needs to be cleanup and clarification. What is now the ActiveSpanManager is merely a 'global' placeholder similar to the 'global' tracer. 
Should it be named GlobalSpanManager again? similar to GlobalTracer??

The `concurrent` package is meant to provide an _example_ for propagating spans accross threads, but by no means is anyone obliged to use it.
I personally prefer an existing context propagation solution to which the SpanManager is registered as callback.
I have this concept working in a library, but it is currently not documented.

The `tracer` package is meant to automatically _activate_ / _deactivate_ spans to be propagated when _started_ / _finished_.
Again, this is not obligatory, but in my opinion a very useful tool to separate the tracing from the propagation.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        public void release() {
+            if (released.compareAndSet(false, true)) {
+                LinkedManagedSpan current = managed.get();
+                if (this == current) {
+                    while (current != null && current.released.get()) {
+                        current = current.parent;
+                    }
+                    if (current == null) managed.remove();
+                    else managed.set(current);
+                    LOGGER.log(Level.FINER, "Deactivated {0} and restored managed span to {1}.", new Object[]{this, current});
```
Comment (wu-sheng) [inline src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java]: All `LOGGER.log()` code should be in the condition judge about whether log or not:
```java
if(LOGGER.isLoggable(Level.FINER)) {
       LOGGER.log(Level.FINER, "Deactivated {0} and restored managed span to {1}.", new Object[] {this, current});
}
```

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        @Override
+        public Span getSpan() {
+            return span;
+        }
+
+        /**
+         * Please see {@link DefaultSpanManager outer class description} for the stack-unwinding documentation.
+         */
+        public void release() {
+            if (released.compareAndSet(false, true)) {
```
Comment (wu-sheng) [inline src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: `DefaultSpanManager` bases on `ThreadLocal` to process in-thread propagation, why need `AtomicBoolean` to avoid concurrency scenario？

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * <li>Otherwise, the first parent that is <em>not yet released</em> is set as the new managed span.</li>
+ * <li>If no managed parents remain, the <em>managed span</em> is cleared.</li>
+ * <li>Consecutive <code>release()</code> calls for already-released spans will be ignored.</li>
+ * </ol>
+ */
+public final class DefaultSpanManager implements SpanManager {
+
+    private static final Logger LOGGER = Logger.getLogger(DefaultSpanManager.class.getName());
+
+    private static final DefaultSpanManager INSTANCE = new DefaultSpanManager();
```
Comment (wu-sheng) [inline src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:39]: As a single instance, with a private construct, and static final field. I prefer the class like this:
```java
public enum DefaultSpanManager implements SpanManager {
    INSTANCE;

    // no need of static field
    // no need of private construct
    // maybe no need of public static SpanManager getInstance()
```

😊, this is just a suggestion. Choose you want.

File: src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    @Override
+    public Span start() {
+        return new AutoReleasingManagedSpan(spanManager.manage(delegate.start()));
+    }
+
+    // All other methods are forwarded to the delegate SpanBuilder.
+
+    public SpanBuilder asChildOf(SpanContext parent) {
+        if (parent instanceof ManagedSpanBuilder) parent = ((ManagedSpanBuilder) parent).delegate;
```
Comment (wu-sheng) [inline src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java]: As I know, `instanceof` is optimized in modern JVM, like JVM8, but I am not sure in JVM6. 
Besides,  if `instanceof` returns false, `delegate` seems not updated, is this what you want?

File: src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanTracer.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    public ManagedSpanTracer(Tracer delegate, SpanManager spanManager) {
+        if (delegate == null) throw new NullPointerException("Delegate Tracer is <null>.");
+        if (spanManager == null) throw new NullPointerException("Span manager is <null>.");
+        this.delegate = delegate;
+        this.spanManager = spanManager;
+    }
+
+    public <C> void inject(SpanContext spanContext, Format<C> format, C carrier) {
+        // Weird that SpanBuilder extends Context!
+        if (spanContext instanceof ManagedSpanBuilder) spanContext = ((ManagedSpanBuilder) spanContext).delegate;
```
Comment (wu-sheng) [inline src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanTracer.java]: Also `instanceof` and `delegate` no-update issue, as I mention before.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 8 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:39]: For a library I prefer the static method as it allows future flexibility.
I agree that in this case it's doubtful it will be needed.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 7 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: It doesn't hurt and actually makes the code simpler by having _if (!released)_ and _released = true;_ statements combined and atomic.
About the concurrency scenario, you bring up a valid point though; maybe the released should _actually_ be a ThreadLocal[Boolean] instead ??
I don't know whether ManagedSpan.release() from another thread than the creator should have well-defined meaning. If it does, I think the behaviour it has now, maybe the simplest option..

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 6 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java]: Do you have measurements to show this is significant (it's the first thing Logger.log does)?
The cyclomatic complexity increases and forces additional testing with various loglevels.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 7 above)
Comment (wu-sheng) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: If use `DefaultSpanManager`, I create a span in thread-A, and stop it in thread-B, is this going to work? 
I am not sure.
The stop in thread-B seems, can't the clear the `LinkedManagedSpan` in thread-A's threadlocal, maybe?

File: src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java (hunk 9 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java]: Delegate is always updated (by rewrap()), I'll add block around the single-line if to make that more clear.
Unfortunately the _instanceof_ is really necessary due to tracers not being forced to accept any context but their own.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 6 above)
Comment (wu-sheng) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java]: Ignore this. My bad.

File: src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java (hunk 9 above)
Comment (wu-sheng) [reply src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java]: As you said, the delegate is updated in `rewrap`, why need `parent = ((ManagedSpanBuilder) parent).delegate`? What is the benefit?
A little confused. 😭

Comment (wu-sheng) [reply src/main/java/io/opentracing/contrib/spanmanager/tracer/ManagedSpanBuilder.java]: `delegate` should be ok as the `asChildOf`'s parameter, that is a delegate meaning.

File: src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java (hunk 7 above)
Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: A span finish() and close() has _no_ relation with the ManagedSpan!
I don't know what you mean by stop().

If you mean ManagedSpan.release(), that is currently undefined behaviour and I don't really know a use-case for it. To _release_ a current Span, it implies it is managed in the current process.
The `Span` is intended to be propagated/managed accross Threads, the `ManagedSpan` is **not**. It is purely intended to restore currentSpan to a previous state.

Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: I think some additional documentation is in order here :smile:

Comment (wu-sheng) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: I see the usage in `RunnableWithManagedSpan`. It's fine. 
Documents are enough.😊

Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: Actually, I improved the predictability by moving the stack unwinding to the manager itself.
Now it should no longer matter from which thread the ManagedSpan is released. If it is 'current' in any thread, it should become 'no longer current' in that thread when released.

Comment (sjoerdtalsma) [reply src/main/java/io/opentracing/contrib/spanmanager/DefaultSpanManager.java:111]: Now adding unit tests for stack unwinding.
Will also add unit testcase for release-from-other-thread (although a far-fetched use-case, I think it should be predictable).

**Comment by sjoerdtalsma** [comment]
@bensigelman Could you add me as a Collaborator to this repo?
Then I can enable Travis builds for it.

**Comment by bhs** [comment]
@sjoerdtalsma can you summarize where we are at with this PR? I see that you changed the README to reflect its early stage... are we just waiting for the release artifact machinery to stabilize?

Thanks!

**Comment by sjoerdtalsma** [comment]
@bensigelman 

- travis needs to be enabled on this repository. I would have done it, but I don't think I'm a Collaborator on the repo (I could enable it on the globaltracer, not on this one).
  >@bensigelman Could you add me as a Collaborator to this repo?
   Then I can enable Travis builds for it.
- the repo needs to be renamed to java-spanmanager before we can finalize on the release packaging in bintray (Don't know if I can rename the repo with Collaborator rights, but otherwise, you could do it).

**Comment by bhs** [comment]
@sjoerdtalsma I saw your previous note and gave you write access to the repo...

**Comment by bhs** [comment]
... I bumped you up to Admin, that should hopefully be sufficient!

image

**Comment by sjoerdtalsma** [comment]
>... I bumped you up to Admin, that should hopefully be sufficient!

That did it! I enabled travis for this repo, thanks!

**Comment by sjoerdtalsma** [comment]
The repository rename needs to happen before release, since all credentials are encrypted to a specific org/repo for travis.
I think then we're good to go. As I'm admin now, I guess I can do the rename too...
Okay with you @bensigelman  ?

**Comment by bhs** [comment]
thanks @sjoerdtalsma!!
